File: 2020/js/day03.js
const read = require('./read.js')
const fn = read('03')

const n = fn[0].length

const traverse = (slop, odd = false) => {
  const res = fn.reduce((acc, s, index) => {
    if (odd && (index % 2 == 1)) return acc
    acc.c += s[acc.i] === '#' ? 1 : 0
    acc.i = (acc.i + slop) % n
    return acc
  }, { i: 0, c: 0 })
  return res.c
}

const part1 = () => traverse(3)
const part2 = () => {
  return [
    traverse(1), traverse(3),
    traverse(5), traverse(7),
    traverse(1, true)
  ]
}

console.log(part1(), part2())
console.log(104*230*83*98*49)

File: 2020/hs/day03.hs
part :: Int -> Int -> [String] -> Int
part dx dy mat = length [s | s <- visits, s == '#']
	where 
		n = length mat
		m = length $ head mat
		pos = [(i, (i*dx `div` dy) `mod` m) | i <- [0,dy..(n-1)]]
		visits = [mat !! (fst p) !! (snd p) | p <- pos]

part1 :: [String] -> Int
part1 = part 3 1

part2 :: [String] -> Int
part2 mat = n1*n3*n5*n7*n2
	where 
		n1 = part 1 1 mat
		n3 = part 3 1 mat
		n5 = part 5 1 mat
		n7 = part 7 1 mat
		n2 = part 1 2 mat

main :: IO()
main = do
	fn <- lines <$> readFile "../res/03.input"
	print $ (part1 fn, part2 fn)